test/jdk/java/io/Reader/Of.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadDirectCharBuffer(Reader reader) throws IOException {
+        CharBuffer charBuffer = ByteBuffer.allocateDirect(32).asCharBuffer();
+        charBuffer.position(8);
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect position or limit");
+        charBuffer.rewind();
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect end of stream");
+        charBuffer.rewind();
+        assertEquals(charBuffer.toString(),
+                // last part first proofs that copy loops correctly stopped
+                CONTENT.substring(8, 16) + CONTENT.substring(0, 8),
+                "read(CharBuffer) provides wrong content");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadNonDirectCharBuffer(Reader reader) throws IOException {
+        CharBuffer charBuffer = CharBuffer.allocate(16);
+        charBuffer.position(8);
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect position or limit");
+        charBuffer.rewind();
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect end of stream");
+        charBuffer.rewind();
+        assertEquals(charBuffer.toString(),
+                CONTENT.substring(8, 16) + CONTENT.substring(0, 8),
+                "read(CharBuffer) provides wrong content");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadCharBufferZeroRemaining(Reader reader) throws IOException {
+        CharBuffer charBuffer = CharBuffer.allocate(0);
+        assertEquals(reader.read(charBuffer), 0, "read(CharBuffer) != 0");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReady(Reader reader) throws IOException {
+        assertTrue(reader.ready());
+    }
+
+    @Test(dataProvider = "readers")
+    public void testSkip(Reader reader) throws IOException {
+        assertEquals(reader.skip(8), 8, "skip() does not respect limit");
+        assertEquals(reader.skip(9), 8, "skip() does not respect end of stream");
+        assertEquals(reader.skip(1), 0, "skip() does not respect empty stream");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testTransferTo(Reader reader) throws IOException {
+        StringWriter sw = new StringWriter(16);
+        assertEquals(reader.transferTo(sw), 16, "transferTo() != 16");
+        assertEquals(reader.transferTo(sw), 0,
+                "transferTo() does not respect empty stream");
+        assertEquals(sw.toString(), CONTENT,
+                "transferTo() provides wrong content");
+    }
+
+    @Test(dataProvider = "readers", expectedExceptions = IOException.class)
+    public void testReadClosed(Reader reader) throws IOException {
+        reader.close();
+        reader.read();
+    }
+
+    @Test(dataProvider = "readers", expectedExceptions = IOException.class)
```

Comment:
Same comment as above, about using assertThrows instead of expectedExceptions at the method level.